### ISSUE TITLE
NativeAOT Library Test Failures Don't Fail CI

### ISSUE THREAD

**Issue by Beau-Gosse-dev** [dotnet/runtime#70223]
Test failures inside the library tests for Native AOT don't show as failed in the CI pipeline even when a test fails.

Example (all checks/jobs/steps below show up as green):
This PR: https://github.com/dotnet/runtime/pull/70173/checks?check_run_id=6716857093

Under Checks->.NET Helix->Build Analysis->Runtime brings you to here: https://dev.azure.com/dnceng/public/_build/results?buildId=1803448&view=results

Find Job "Build windows arm64 Release NativeAOT" with step "Send to Helix" which shows: https://dev.azure.com/dnceng/public/_build/results?buildId=1803448&view=logs&j=7bb91cf6-d57c-5393-cf5c-c884f1e50962&t=7729b4d4-9255-54d8-01fd-e831ebbb744b

From there, you can see "Sent Helix Job; see work items at https://helix.dot.net/api/jobs/7528557f-cbe9-4665-8fa3-6028facd5084/workitems?api-version=2019-06-17" 

Follow that link and you will see the "System.Reflection.Tests" details URL: https://helix.dot.net/api/jobs/7528557f-cbe9-4665-8fa3-6028facd5084/workitems/System.Reflection.Tests?api-version=2019-06-17

And from there, you can see the "ConsoleOutputUri  https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-70173-merge-7528557fcbe946658f/System.Reflection.Tests/1/console.ba43aa3a.log?helixlogtype=result"

Which shows this:
```
System.PlatformNotSupportedException : Operation is not supported on this platform.
   at System.Reflection.Assembly.GetCallingAssembly() + 0x6c
   at System.Reflection.TestAssembly.ClassToInvoke.CallMeActual() + 0xc
   at System.Reflection.TestAssembly.ClassToInvoke.CallMe_AggressiveInlining() + 0x10
   at System.Reflection.Tests!<BaseAddress>+0x681698
   at System.InvokeUtils.CallDynamicInvokeMethod(Object, IntPtr, IntPtr, IntPtr, Object, Object[], BinderBundle, Boolean, Boolean) + 0xec
```

Running these tests manually locally also results in this error.

**Comment by Beau-Gosse-dev** [comment]
@MichalStrehovsky and @LakshanF  FYI

**Comment by MichalStrehovsky** [comment]
@LakshanF  we found this in https://github.com/dotnet/runtime/pull/69132#issuecomment-1144368841 - did you have a chance to look at this? This is higher priority than any other library test work because we can't see regressions even in the things we already made working.

**Comment by Beau-Gosse-dev** [comment]
I don't know anything about Helix (what is Helix?) But I thought this line was interesting. Specifically "category=IgnoreForCI" and "category=failing" 

`System.Reflection.Tests.exe -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing `

From here: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-70167-merge-8ff4a8b780fc4dceb6/System.Reflection.Tests/1/console.397d8933.log?helixlogtype=result

**Comment by MichalStrehovsky** [comment]
What's helix is described in the glossary: https://github.com/dotnet/runtime/blob/45589f2a1ea20b4bce66058994de52f317716558/docs/project/glossary.md

The notrait arguments just mean that we don't want to look at tests that are marked failing (eg ActiveIssue), outerloop, or ignoreforci. The arguments are expected.

**Comment by LakshanF** [comment]
> @LakshanF we found this in [#69132 (comment)](https://github.com/dotnet/runtime/pull/69132#issuecomment-1144368841) - did you have a chance to look at this? This is higher priority than any other library test work because we can't see regressions even in the things we already made working.

Taking a look

**Comment by LakshanF** [comment]
Comparing NativeAot 'passing' [job](https://dev.azure.com/dnceng/public/_build/results?buildId=1802634&view=logs&j=7bb91cf6-d57c-5393-cf5c-c884f1e50962&t=7729b4d4-9255-54d8-01fd-e831ebbb744b) with a baseline job, "Send to Helix", that was intentionally [failed](https://dev.azure.com/dnceng/public/_build/results?buildId=1806874&view=logs&j=55b384d0-d6b0-5894-fd8a-d754d713246e&t=8dcacd81-cb5e-5ac9-93e5-c3f19b19e19b), the following steps are common
 - Building Helix work items
 - Sending Job to _Helix Machine_
 - Waiting for job to finish
 - Job Completed with org work items + 1
 - Stopping the test run at _Helix Machine_

The baseline failing run has an additional check from microsoft.dotnet.helix.sdk package that seems to look at the results from the test run which isn't present in the 'passing' NativeAOT job. 

`D:\a\_work\1\s\.packages\microsoft.dotnet.helix.sdk\7.0.0-beta.22266.1\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(43,5): error : Test System.Collections.Generic.Tests.DecimalComparersTests.EqualityComparer_GetHashCodeShouldBeOverriddenAndBeTheSameAsLongAsTheTypeIsTheSame_cloned has failed. Check the ...`

Not sure if this is relevant and how this step gets enabled. Need to talk to someone who is familiar (@jkoritzinsky or @MattGal)  with the process.

**Comment by jkoritzinsky** [comment]
I think the test results file isn’t being written out for some reason, or is corrupted. The System.Linq.Expressions.Tests work item console log says that no test result files could be found.

**Comment by LakshanF** [comment]
@jkoritzinsky, the test runner for this is, [SingleTestRunner](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/SingleFileTestRunner/SingleFileTestRunner.cs),  and I wonder if this is not hooking up to the correct way to write results.

Another possibility is the one @MattGal mentioned that this job is not set up to fail if there are test failures (similar to staging tests) but I couldn't see an obvious difference with [this step ](https://github.com/dotnet/runtime/blob/main/eng/pipelines/runtime.yml#L247-L274)in runtime.yml and others.

**Comment by jkoritzinsky** [comment]
I think the SingleTestRunner needs to write out the `resultsXmlAssembly` object to disk as a `testResults.xml` file to actually report results. It looks like it just collects the test results into the object but never writes it out.

**Comment by MichalStrehovsky** [comment]
Just to add to what Jeremy wrote: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-70238-merge-d78e61af15cb45b1af/System.Collections.Tests/1/console.8810c498.log?helixlogtype=result indicates that Helix is passing `-xml testResults.xml` argument to the test. We might just need to steal the processing of that argument from the "real" runner: https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.XUnitConsoleRunner/src

When single file test runner was originally written, I pointed out that the "real" runner used to support "single-file-like" scenarios for .NET Native on UWP but that support was gutted in https://github.com/dotnet/arcade/pull/4142 (and maybe others) when we were getting rid of .NET Native/UWP, but we went with this new runner instead.

Maybe medium-to-longer term (not _now_) we should try to get rid of the special-unicorn-single-file runner in favor of making the real runner work for single file. It will avoid compatibility issues like this, or https://github.com/dotnet/runtimelab/issues/901.
